how-to/troubleshoot: add tidb diagnose map (#2440)
pingcap/docs: how-to/troubleshoot/diagnose-map.md

---
title: TiDB Troubleshooting Map
summary: Learns how to troubleshoot common issues in TiDB.
category: how-to
---

# TiDB Troubleshooting Map

## 1. Service Unavailable

### 1.1 The client reports `Region is Unavailable` error

> - 1.1.1 `Region is Unavailable` is usually because a Region is not available for a period of time. You might encounter `TiKV server is busy`, or the request to TiKV fails due to `not leader` or `epoch not match`，or the request to TiKV time out. In such cases, TiDB performs a `backoff` retry mechanism. When the `backoff` exceeds a threshold (20s by default), the error will be sent to the client. Within the `backoff` threshold, this errors is not visible to the client.

Reviewer: ```suggestion
- 1.1.1 `Region is Unavailable` is usually because a Region is not available for a period of time. You might encounter `TiKV server is busy`, or the request to TiKV fails due to `not leader` or `epoch not match`, or the request to TiKV time out. In such cases, TiDB performs a `backoff` retry mechanism. When the `backoff` exceeds a threshold (20s by default), the error will be sent to the client. Within the `backoff` threshold, this errors is not visible to the client.
```

> - 1.1.2 Multiple TiKV instances OOM at the same time, which causes no Leader in a Region for a period of time. See [case-991](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case991.md) in Chinese.

Reviewer: ```suggestion
- 1.1.2 Multiple TiKV instances are OOM at the same time, which causes no Leader in a Region for a period of time. See [case-991](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case991.md) in Chinese.
```

> - 1.1.3 TiKV reports `TiKV server is busy`，and exceeds the `backoff` time. For more details, refer to 4.3. `TiKV server is busy` is a result of the internal flow control mechanism and should not be counted in the `backoff` time. This issue will be fixed.

Reviewer: ```suggestion
- 1.1.3 TiKV reports `TiKV server is busy`, and exceeds the `backoff` time. For more details, refer to [4.3](#43-the-client-reports-the-server-is-busy-error). `TiKV server is busy` is a result of the internal flow control mechanism and should not be counted in the `backoff` time. This issue will be fixed.
```

> - 1.1.4 Multiple TiKV instances failed to start, which causes no Leader in a Region. When multiple TiKV instances are deployed in a physical machine, the failure of the physical machine can cause no Leader in a Region if the label is not property configured. See [case-228](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case228.md) in Chinese.

Reviewer: ```suggestion
- 1.1.4 Multiple TiKV instances failed to start, which causes no Leader in a Region. When multiple TiKV instances are deployed in a physical machine, the failure of the physical machine can cause no Leader in a Region if the label is not properly configured. See [case-228](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case228.md) in Chinese.
```

- 1.1.5 When a follower apply is lagged in a previous epoch, after the Follower becomes Leader, it rejects the request with `epoch not match`. See [case-958](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case958.md) in Chinese (TiKV needs to optimize its mechanism).

### 1.2 PD errors cause service unavailable

> Refer to 5 PD issues.

Reviewer: Suggest adding links to make it more friendly.
```suggestion
Refer to [PD issues](#5-pd-issues).
```

## 2. Latency increases significantly

### 2.1 Transient increase

- 2.1.1 Wrong TiDB execution plan causes latency increase. Refer to 3.3.
- 2.1.2 PD Leader election issue or OOM. Refer to 5.2 and 5.3.
- 2.1.3 A significant number of Leader drops in some TiKV instances. Refer to 4.4.

### 2.2 Persistent and significant increase

- 2.2.1 TiKV single thread bottleneck

>     - Too many regions in a TiKV instance causes a single gRPC thread to be the bottleneck (Check the Grafana -> TiKV-details -> `Thread CPU/gRPC CPU Per Thread` metric).In versions above v3.x, you can enable `Hibernate Region` to resolve the issue. See [case-612](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case612.md) in Chinese.

Reviewer: ```suggestion
    - Too many Regions in a TiKV instance causes a single gRPC thread to be the bottleneck (Check the Grafana -> TiKV-details -> `Thread CPU/gRPC CPU Per Thread` metric). In v3.x or later versions, you can enable `Hibernate Region` to resolve the issue. See [case-612](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case612.md) in Chinese.
```

>     - For versions earlier than v3.0, when the raftstore thread or the apply thread becomes the bottleneck (Grafana -> TiKV-details -> `Thread CPU/raft store CPU` and `Async apply CPU` exceeds `80%`), you can scale out TiKV (v2.x)instances or upgrade to v3.x with multi-threading. <!-- See [case-517](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case517.md) in Chinese. -->

Reviewer: ```suggestion
    - For versions earlier than v3.0, when the raftstore thread or the apply thread becomes the bottleneck (Grafana -> TiKV-details -> `Thread CPU/raft store CPU` and `Async apply CPU` exceeds `80%`), you can scale out TiKV (v2.x) instances or upgrade to v3.x with multi-threading. <!-- See [case-517](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case517.md) in Chinese. -->
```

- 2.2.2 CPU load increase

- 2.2.3 TiKV slow write. Refer to 4.5

- 2.2.4 TiDB wrong execution plan. Refer to 3.3

## 3. TiDB issues

### 3.1 DDL

> - 3.1.1  An error `ERROR 1105 (HY000): unsupported modify decimal column precision` is reported when you modify the length of `decimal` field.<!--See [case-1004](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case517.md) in Chinese.--> TiDB does not support changing the length of `decimal` field.

Reviewer: ```suggestion
- 3.1.1  An error `ERROR 1105 (HY000): unsupported modify decimal column precision` is reported when you modify the length of the `decimal` field.<!--See [case-1004](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case517.md) in Chinese.--> TiDB does not support changing the length of the `decimal` field.
```

- 3.1.2 TiDB DDL job hangs or executes slowly (use `admin show ddl jobs` to check DDL progress)

>     - Cause 1：Network issue to other components（PD / TiKV).

Reviewer: ```suggestion
    - Cause 1：Network issue to other components (PD/TiKV).
```

    - Cause 2：Early versions of TiDB (earlier than v3.0.8) have heavy internal load because of a lot of goroutine at high concurrency.

    - Cause 3：In early versions (v2.1.15 & version < v3.0.0-rc1), PD instances fail to delete TiDB keys, which causes every DDL change to wait for two leases.

    - For other unknown causes, file issues at [PingCAP/TiDB](https://github.com/pingcap/tidb).

    - Solution:

        - For cause 1, check the network connection between TiDB and TiKV/PD. 

>         - For cause 2 and 3, the issues are already fixed in later versions. You can upgrade to a higher version. 

Reviewer: ```suggestion
        - For cause 2 and 3, the issues are already fixed in later versions. You can upgrade TiDB to a later version. 
```

>         - For other causes, you can use the following plan for DDL owner migration.

Reviewer: ```suggestion
        - For other causes, you can use the following solution of migrating the DDL owner.
```

    - DDL owner migration:

>         - If you can connect to the TiDB server, execute the owner election command again：`curl -X POST http://{TiDBIP}:10080/ddl/owner/resign`

Reviewer: ```suggestion
        - If you can connect to the TiDB server, execute the owner election command again: `curl -X POST http://{TiDBIP}:10080/ddl/owner/resign`
```

>         - 2：if you cannot connect to the TiDB server, use `tidb-ctl` to delete the DDL owner from the etcd of the PD cluster to trigger re-election：`tidb-ctl etcd delowner [LeaseID] [flags] + ownerKey`

Reviewer: ```suggestion
        - If you cannot connect to the TiDB server, use `tidb-ctl` to delete the DDL owner from the etcd of the PD cluster to trigger re-election: `tidb-ctl etcd delowner [LeaseID] [flags] + ownerKey`
```

- 3.1.3 TiDB reports `information schema is changed` error in log

    - Cause 1: The DML operation touches a table that is under DDL. You can use `admin show ddl job` to check the DDLs that are currently in progress.

    - Cause 2: The current DML executes too long. During the time, many DDLs are executed, which causes `schema version` changes to be more than 1024. The new version `lock table` might also cause schema version changes.

>     - Cause 3: The TiDB instance that are currently executing DML cannot load the new `schema information` (maybe caused by network issues with PD or TiKV). During this time, many DDL statements are executed (including `lock table`), which causes `schema version` changes to be more than 1024.

Reviewer: ```suggestion
    - Cause 3: The TiDB instance that is currently executing DML statements cannot load the new `schema information` (maybe caused by network issues with PD or TiKV). During this time, many DDL statements are executed (including `lock table`), which causes `schema version` changes to be more than 1024.
```

>     - Solution：The first two causes do not impact the application, as the related DMLs retry after the failure. For cause 3, you need to check the network between TiDB and TiKV/PD.

Reviewer: ```suggestion
    - Solution：The first two causes do not impact the application, as the related DML operations retry after failure. For cause 3, you need to check the network between TiDB and TiKV/PD.
```

>     - Background: The increase number of `schema version` is consistent with the number of `schema state` of each DDL change operation. For example, the `create table` operation has 1 version change, and the `add column` operation has 4 version changes. Therefore, too many column change operations might cause `schema version` to increase fast. For details, refer to [online schema change](https://static.googleusercontent.com/media/research.google.com/zh-CN//pubs/archive/41376.pdf). 

Reviewer: ```suggestion
    - Background: The increased number of `schema version` is consistent with the number of `schema state` of each DDL change operation. For example, the `create table` operation has 1 version change, and the `add column` operation has 4 version changes. Therefore, too many column change operations might cause `schema version` to increase fast. For details, refer to [online schema change](https://static.googleusercontent.com/media/research.google.com/zh-CN//pubs/archive/41376.pdf).
```

        - The TiDB server did not load the new schema within one DDL lease (`45s` by default); or
        - The TiDB server disconnects with PD for connections with `keep alive`. 

    - Cause 3: TiKV has high load or network timed out. Check the node loads in `Grafana -> TiDB and TiKV`.

    - Solution:
    
        - For cause 1, retry the DML when TiDB is started.
        - For cause 2, check the network between the TiDB server and PD/TiKV.
        - For cause 3, investigate why TiKV is busy. Refer to 4 TiKV issues.

### 3.2 OOM issues

- 3.2.1 Symptom

    - Client：The client reports error `ERROR 2013 (HY000): Lost connection to MySQL server during query`.

    - Check the log

        - Execute `dmesg -T | grep tidb-server`. The result shows the OOM-killer log around the time point when the error occurs.

        - Grep the "Welcome to TiDB" log around the time point after the error occurs (namely, the time when tidb-server restarts).

        - Grep `fatal error: runtime: out of memory` or `cannot allocate memory` in `tidb_stderr.log`.

        - In v2.1.8 or earlier versions, you can grep `fatal error: stack overflow` in the `tidb_stderr.log`.

>     - Monitor：The memory usage of tidb-server instance increases sharply in a short period of time. 

Reviewer: ```suggestion
    - Monitor：The memory usage of tidb-server instances increases sharply in a short period of time. 
```

> - 3.2.2 Locate the SQL that causes OOM. (Currently all versions cannot locate SQL accurately. You need to analyze whether OOM is caused by the SQL after you find the SQL.)

Reviewer: ```suggestion
- 3.2.2 Locate the SQL statement that causes OOM. (Currently all versions of TiDB cannot locate SQL accurately. You still need to analyze whether OOM is caused by the SQL statement after you locate one.)
```

- 3.3.3 Mitigation

    - For v3.0 and later versions, use the `SQL Bind` feature to bind the execution plan.
    
    - Update the statistics. If you are roughly sure that the problem is caused by the statistics, [dump the statistics](https://pingcap.com/docs/stable/reference/performance/statistics/#export-statistics). If the cause is outdated statistics, such as the `modify count/row count` in `show stats_meta` is greater than a certain value (e.g. 0.3), or the table has an index of time column, you can try recovering by using `analyze table`. If `auto analyze` is configured, check whether the `tidb_auto_analyze_ratio` system variable is too large (e.g. > 0.3), and whether the current time is between `tidb_auto_analyze_start_time` and `tidb_auto_analyze_end_time`.

    - For other situations, [report a bug](https://github.com/pingcap/tidb/issues/new?labels=type%2Fbug&template=bug-report.md).

### 3.4 SQL execution error

> - 3.4.1 The client reports the `ERROR 1265(01000) Data Truncated` error. This is because the way TiDB internally calculates the precision of `Decimal` type is incompatible with that of MySQL (expected to be fixed in v3.0.10).

Author: Confirmed. This issue is fixed in v3.0.10 (https://github.com/pingcap/tidb/pull/14438).

        - If the default CF compaction sees a high pressure, change the `[rocksdb.defaultcf] compression-per-level` parameter from `["no", "no", "lz4", "lz4", "lz4", "zstd", "zstd"]` to `["no", "no", "zstd", "zstd", "zstd", "zstd", "zstd"]`.

    - Too many memtables causes stall. This usually occurs when the amount of instant writes is large and the memtables flush to the disk slowly. If the disk write speed cannot be improved, and this issue only occurs during business peaks, you can mitigate it by increasing the `max-write-buffer-number` of the corresponding CF.

        - For example, set `[rocksdb.defaultcf] max-write-buffer-number` to `8` (`5` by default). Note that this might cause more memory usage in the peak, because more memtables might be in the memory.

- 4.3.2 `scheduler too busy`

>     - Serious write conflict. `latch wait duration` is high. You can view `latch wait duration` in the monitor **Grafana** -> **TiKV-details** -> **scheduler prewrite**/**scheduler commit**. When the write tasks pile up in the scheduler, the pending write tasks exceed the threshold set in `[storage] scheduler-pending-write-threshold` (100MB). TODO: Verify the cause by viewing the metric corresponding to `MVCC_CONFLICT_COUNTER`.

Reviewer: ```suggestion
    - Serious write conflict. `latch wait duration` is high. You can view `latch wait duration` in the monitor **Grafana** -> **TiKV-details** -> **scheduler prewrite**/**scheduler commit**. When the write tasks pile up in the scheduler, the pending write tasks exceed the threshold set in `[storage] scheduler-pending-write-threshold` (100MB). You can verify the cause by viewing the metric corresponding to `MVCC_CONFLICT_COUNTER`.
```

- 4.3.3 `raftstore is busy`. The processing of messages is slower than the receiving of messages. The short-term `channel full` status does not affect the service, but if the error persists for a long time, it might cause Leader switch.

    - `append log` encounters stall. Refer to [4.3.1](#43-the-client-reports-the-server-is-busy-error).
    - `append log duration` is high, which causes slow processing of messages. You can refer to [4.5](#45-tikv-write-is-slow) to analyze why `append log duration` is high.
    - raftstore receives a large batch of messages in an instant (check in the TiKV Raft messages dashboard), and fails to process them. Usually the short-term `channel full` status does not affect the service.

- 4.3.4 TiKV coprocessor is in a queue. The number of piled up tasks exceeds `coprocessor threads * readpool.coprocessor.max-tasks-per-worker-[normal|low|high]`. Too many large queries leads to the tasks piling up in coprocessor. You need to check whether a execution plan change causes a large number of table scan operations. Refer to [3.3](#33-wrong-execution-plan).

### 4.4 Some TiKV nodes drop Leader frequently

- 4.4.1 Re-election because TiKV is restarted

>     - After TiKV panics, it is pulled up by systemd and runs normally. You can check whether panic has occurred by viewing the TiKV log. Since this issue is unexpected, [report a bug](https://github.com/tikv/tikv/issues/new?template=bug-report.md) if it happens.

Reviewer: ```suggestion
    - After TiKV panics, it is pulled up by systemd and runs normally. You can check whether panic has occurred by viewing the TiKV log. Because this issue is unexpected, [report a bug](https://github.com/tikv/tikv/issues/new?template=bug-report.md) if it happens.
```

    
    - TiKV is stopped or killed by a third party and then pulled up by systemd. Check the cause by viewing `dmesg` and the TiKV log. 
    
    - TiKV is OOM, which causes restart. Refer to [4.2](#42-tikv-oom).
    

>     - TiKV is hung because of dynamic adjust of `THP` (Transparent Hugepage). See case [case-500](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case500.md) in Chinese.

Reviewer: ```suggestion
    - TiKV is hung because of dynamically adjusting `THP` (Transparent Hugepage). See case [case-500](https://github.com/pingcap/tidb-map/blob/master/maps/diagnose-case-study/case500.md) in Chinese.
```

> - 4.4.2 TiKV RocksDB encounters write stall and thus results in re-election. You chan check if the monitor **Grafana** -> **TiKV-details** -> **errors** shows `server is busy`. Refer to [4.3.1](#43-the-client-reports-the-server-is-busy-error).

Reviewer: ```suggestion
- 4.4.2 TiKV RocksDB encounters write stall and thus results in re-election. You can check if the monitor **Grafana** -> **TiKV-details** -> **errors** shows `server is busy`. Refer to [4.3.1](#43-the-client-reports-the-server-is-busy-error).
```

- 4.4.3 Re-election because of network isolation.

### 4.5 TiKV write is slow

> - 4.5.1 Check whether the TiKV write is low by viewing the `prewrite/commit/raw-put` duration of TiKV gRPC (only for raw kv clusters). Generally, you can locate the slow phase according to the [performance-map](https://github.com/pingcap/tidb-map/blob/master/maps/performance-map.png). Some common situations are listed as follows.

Reviewer: ```suggestion
- 4.5.1 Check whether the TiKV write is low by viewing the `prewrite/commit/raw-put` duration of TiKV gRPC (only for raw KV clusters). Generally, you can locate the slow phase according to the [performance-map](https://github.com/pingcap/tidb-map/blob/master/maps/performance-map.png). Some common situations are listed as follows.
```

- 4.5.2 The scheduler CPU is busy (only for transaction kv). 

    The `scheduler command duration` of prewrite/commit is longer than the sum of `scheduler latch wait duration` and `storage async write duration`. The scheduler worker has a high CPU demand, such as over 80% of `scheduler-worker-pool-size` * 100%, or the CPU resources of the entire machine are relatively limited. If the write workload is large, check if `[storage] scheduler-worker-pool-size` is set too small.
    
    For other situations, [report a bug](https://github.com/tikv/tikv/issues/new?template=bug-report.md).

- 4.5.3 Append log is slow.

    The **Raft IO**/`append log duration` in TiKV Grafana is high, usually because the disk write operation is slow. You can verify the cause by checking the `WAL Sync Duration max` value of RocksDB - raft.
    
    For other situations, [report a bug](https://github.com/tikv/tikv/issues/new?template=bug-report.md).

- 4.5.4 The raftstore thread is busy.

>     The Raft Propose/propose wait duration is significantly larger than the append log duration in TiKV Grafana. Take the following methods: 

Reviewer: ```suggestion
    The **Raft Propose**/`propose wait duration` is significantly larger than the append log duration in TiKV Grafana. Take the following methods: 
```

>     - Check whether the `[raftstore] store-pool-size` configuration is too small. It is recommended to set the value between `1` and `5` and not too large.

Reviewer: ```suggestion
    - Check whether the `[raftstore] store-pool-size` configuration value is too small. It is recommended to set the value between `1` and `5` and not too large.
```

- 6.2.2 `Access denied for user 'root'@'172.31.43.27' (using password: YES)` shows when you run `query status` or check the log.

    - The database related passwords in all the DM configuration files must be encrypted by `dmctl`. If a database password is empty, it is unnecessary to encrypt the password.
    - During DM operation, the user of the upstream and downstream databases must have the corresponding read and write privileges. Data Migration also [prechecks the corresponding privileges](https://pingcap.com/docs/tidb-data-migration/stable/precheck/) automatically while starting the data replication task.
    - To deploy different versions of DM-worker/DM-master/dmctl in a DM cluster, see the [case study on AskTUG](https://asktug.com/t/dm1-0-0-ga-access-denied-for-user/1049/5) in Chinese.

- 6.2.3 A replication task is interrupted with the `driver: bad connection` error returned.

    - The `driver: bad connection` error indicates that an anomaly has occurred in the connection between DM and the downstream TiDB database (such as network failure, TiDB restart and so on), and that the data of the current request has not yet been sent to TiDB.

>         - For versions earlier than 1.0.0 GA, stop the task by running `stop-task` and then restart the task by running `start-task`.

Reviewer: Make it clearer?
```suggestion
        - For versions earlier than DM 1.0.0 GA, stop the task by running `stop-task` and then restart the task by running `start-task`.
```

>         - For 1.0.0 GA or later versions, an automatic retry mechanism for this type of error is added. See [#265](https://github.com/pingcap/dm/pull/265).

Reviewer: ```suggestion
        - For DM 1.0.0 GA or later versions, an automatic retry mechanism for this type of error is added. See [#265](https://github.com/pingcap/dm/pull/265).
```

- 6.2.7 The DM replication process returns an error `Error 1366: incorrect utf8 value eda0bdedb29d(\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd)`.

    - This value cannot be successfully written into MySQL 8.0 or TiDB, but can be written into MySQL 5.7. You can skip the data format check by enabling the `tidb_skip_utf8_check` parameter.

### 6.3 TiDB Lightning

- 6.3.1 TiDB Lightning is a tool for fast full import of large amounts of data into a TiDB cluster. See [TiDB Lightning on GitHub](https://github.com/pingcap/tidb-lightning).

- 6.3.2 Import speed is too slow.

    - `region-concurrency` is set too high, which causes thread contention and reduces performance. Three ways to troubleshoot:

        - The setting can be found from the start of the log by searching `region-concurrency`.
        - If TiDB Lightning shares a server with other services (e.g. Importer), you must manually set `region-concurrency` to 75% of the total number of CPU cores on that server.

>         - If there is a quota on CPU (e.g. limited by Kubernetes settings), Lightning might not be able to read this out. In this case, `region-concurrency` must also be manually reduced.

Reviewer: ```suggestion
        - If there is a quota on CPU (for example, limited by Kubernetes settings), TiDB Lightning might not be able to read this out. In this case, `region-concurrency` must also be manually reduced.
```
